File: javascript/0953-verifying-an-alien-dictionary.js
var isAlienSorted = function(words, order) {
    // first differing char
    // if word A is prefix of word B, word B must be AFTER word A
    orderInd = new Map(); {
        let ind = 0;
        for(const c of order)
            orderInd.set(c, ind++);
    }
    
    for(let i = 0; i < words.length - 1; i++) {
        let w1 = words[i], w2 = words[i + 1];
        
        for(let j = 0; j < w1.length; j++) {
            if(j == w2.length)
                return false;
            
            if(w1.charAt(j) != w2.charAt(j))
                if(orderInd.get(w2.charAt(j)) < orderInd.get(w1.charAt(j)))
                    return false;
                else
                    break;
        }
    }
    return true;
};

File: go/0953-verifying-an-alien-dictionary.go
func isAlienSorted(words []string, order string) bool {
    // first differing char
    // if word A is prefix of word B, word B must be AFTER word A
    orderInd := make(map[byte]int)
    for i := 0; i < len(order); i++ {
        orderInd[order[i]] = i
    }
    
    for i := 0; i < len(words) - 1; i++ {
        w1, w2 := words[i], words[i + 1]
        
        for j := 0; j < len(w1); j++ {
            if j == len(w2) {
                return false
            }
            
            if w1[j] != w2[j] {
                if orderInd[w2[j]] < orderInd[w1[j]] {
                    return false
                }
                break
            }
        }
    }
    return true
}
